[NNC] Dont inlinine intermediates w/multiple uses on cpu

eellison opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#49565 [NNC] Dont inlinine intermediates w/multiple uses on cpu**

Differential Revision: [D25688271](https://our.internmc.facebook.com/intern/diff/D25688271)

Commit:
Dont inlinine intermediates on cpu

[ghstack-poisoned]

ZolotukhinM reviewed (commented):
I think that's too big of a hammer. During the lowering we create lots of intermediates, I suspect it could be very inefficient if we don't inline any of them. As one simple heuristic what to inline, we could inline intermediates with a single use.

This is for instance our original stmt for
```
def foo(x):
  return x*x*x
```
Stmt:
```
[DEBUG kernel.cpp:1488] Original Stmt:
[DEBUG kernel.cpp:1488] {
[DEBUG kernel.cpp:1488]   for (int i0 = 0; i0 < 256; i0++) {
[DEBUG kernel.cpp:1488]     input1[i0] = t0[0 + i0 * 1];
[DEBUG kernel.cpp:1488]   }
[DEBUG kernel.cpp:1488]   for (int v = 0; v < 256; v++) {
[DEBUG kernel.cpp:1488]     aten_mul[v] = (input1(v)) * (input1(v));
[DEBUG kernel.cpp:1488]   }
[DEBUG kernel.cpp:1488]   for (int v_1 = 0; v_1 < 256; v_1++) {
[DEBUG kernel.cpp:1488]     aten_mul_1[v_1] = (aten_mul(v_1)) * (input1(v_1));
[DEBUG kernel.cpp:1488]   }
```

eellison commented:
Doesn't compute inline still handle that ?

ZolotukhinM commented:
> I thought compute inline handled that ?

I think I might have lost your thought. Doesn't this change disable compute inline for CPU?

eellison commented:
I don't think so. I ran 
```
      graph(%0 : Float(5, 3, strides=[3, 1], device=${device}),
            %1 : Float(5, 3, strides=[3, 1], device=${device})):
        %2 : Float(5, 3, strides=[3, 1]) = aten::mul(%0, %1)
        %one : int = prim::Constant[value=1]()
        %4 : Float(5, 3, strides=[3, 1]) = aten::add(%0, %2, %one)
        %5: Float(5, 3, strides=[3, 1]) = aten::add(%4, %2, %one)
        return (%5))IR";
```

And the generated code:
```
{
  if (512 * blockIdx.x + threadIdx.x<15 ? 1 : 0) {
    float t0_1 = t0[512 * blockIdx.x + threadIdx.x]; 
    float t1_1 = t1[512 * blockIdx.x + threadIdx.x]; 
    aten_add[512 * blockIdx.x + threadIdx.x] = (t0_1 + t0_1 * t1_1) + t0_1 * t1_1;
  }
}
```

I  think `computeInline` is still being run, separate from `inlineIntermediates`.

eellison commented:
@ZolotukhinM edit, yea you're right

Commit:
Update on "Dont inlinine intermediates on cpu"




[ghstack-poisoned]

eellison commented:
Okay, so update to previous debugging statements:

It's true that `ir_visitor` and `ir_mutator` don't visit the `buf` contained in `FunctionCall`, however that doesn't actually affect `UseFinder`, because `UseFinder` only looks for loads and stores, not any use at all. I looked into making `UseFinder` also look at FunctionCalls, but that didn't work because UseFinder only is looking for stmts, which have access to their parent block, and `FunctionCall` is an expr. So, instead I updated the name of UseFinder to be more accurate. 

Separately, memDependencyChecker maybe is only supposed to be used after FunctionCalls have been lowered from the IR, it would be nice to throw if they are present or does not work with intermediate buffers, i'm not sure. 

Either way - this fixes a blocker for enabling cpu fusion,  I would like to land this first and then afterwards see if UseFinder / FunctionCallUses etc should be merged.

Commit:
Update on "Dont inlinine intermediates on cpu"


Differential Revision: [D25688271](https://our.internmc.facebook.com/intern/diff/D25688271)

[ghstack-poisoned]

Commit:
Update on "Dont inlinine intermediates on cpu"


Differential Revision: [D25688271](https://our.internmc.facebook.com/intern/diff/D25688271)

[ghstack-poisoned]

ZolotukhinM commented on torch/csrc/jit/tensorexpr/loopnest.cpp:
```diff
@@ -751,28 +777,50 @@ bool LoopNest::computeInline(const Buf* b) {
-  // cpu code generation but is enabled on gpu because it avoids difficult
-  // synchronization logic across blocks.
-  if (inline_output_buffers) {
+  if (allow_duplicated_work) {
```
Nit: might be worth mentioning that inlining a statement with multiple uses doesn't *always* results in a duplicated work: trivial copies (a[i] = b[i]) can be inlined everywhere and it will not be more expensive than the non-inlined version.

eellison replied:
about that... see comment, what do you think?

ZolotukhinM reviewed (approved):
Looks good! I agree that we should revisit the `FunctionCall` lowering semantics, but this fix seems good for now.

eellison commented:
Hmm, so one other wrinkle:
```
  const auto graph_string = R"IR(
      graph(%0 : Float(5, 3, strides=[3, 1], device=cpu),
            %1 : Float(5, 3, strides=[3, 1], device=cpu)):
        %2 : Float(5, 3, strides=[3, 1]) = aten::mul(%0, %1)
        %3 : Float(5, 3, strides=[3, 1]) = aten::mul(%0, %2)
        return (%3))IR";
```

Here, %0 and %1 get read into an intermediate buffer, and %0 has multiple uses, so the %0 intermediate buffer reads dont get inlined and we get an extra loop that is just initializing input1.  
```
{
  Allocate(input1, float, {15});
  for (int i0 = 0; i0 < 5; i0++) {
    for (int i1_tail_tail = 0; i1_tail_tail < 3; i1_tail_tail++) {
      input1[3 * i0 + i1_tail_tail] = t0[3 * i0 + i1_tail_tail];
    }
  }
  for (int v = 0; v < 5; v++) {
    for (int _tail_tail = 0; _tail_tail < 3; _tail_tail++) {
      aten_mul[3 * v + _tail_tail] = (input1[3 * v + _tail_tail]) * ((input1[3 * v + _tail_tail]) * (t1[3 * v + _tail_tail]));
    }
  }
  Free(input1);
```

This actually is desirable if the intermediate buffer taken is taken from a strided tensor. Maybe there's a way in [bindInput](https://github.com/pytorch/pytorch/blob/master/torch/csrc/jit/tensorexpr/kernel.cpp#L1688) to avoid creating an intermediate for contiguous strided tensors ?

Or, I could look into seeing if an intermediate tensor only has one dependency that's an input tensor and always inline it ?

WDYT @ZolotukhinM @nickgg ?

ZolotukhinM commented:
> Here, %0 and %1 get read into an intermediate buffer, and %0 has multiple uses, so the %0 intermediate buffer reads dont get inlined and we get an extra loop that is just initializing input1.
> This actually is desirable if the intermediate buffer taken is taken from a strided tensor. Maybe there's a way in bindInput to avoid creating an intermediate for contiguous strided tensors ?

It might be a good idea, but I haven't thought much about it. Another thing to consider here is a future support for `aten::view`. It probably could be modeled in a similar way - i.e. the intermediate tensor is just a view into the input tensor and we should avoid materializing it if we can. Just 2¢. For now whatever is easier would do I think.

Commit:
Update on "[NNC] Dont inlinine intermediates w/multiple uses on cpu"


Differential Revision: [D25688271](https://our.internmc.facebook.com/intern/diff/D25688271)

[ghstack-poisoned]

eellison commented:
I commented and deleted bc i figured i would just implement it first and then tag, but you saw my comment too quick 😛 

The test that was breaking was: 
```
  const auto graph_string = R"IR(
      graph(%0 : Float(5, 3, strides=[3, 1], device=cpu),
            %1 : Float(5, 3, strides=[3, 1], device=cpu)):
        %2 : Float(5, 3, strides=[3, 1]) = aten::mul(%0, %1)
        %3 : Float(5, 3, strides=[3, 1]) = aten::mul(%0, %2)
        return (%3))IR";
```

Here, %0 and %1 get read into an intermediate buffer, and %0 has multiple uses, so the %0 intermediate buffer reads dont get inlined and we get an extra loop that is just initializing input1.  

I updated the inlining logic to look for direct reads from input buffers and inline them.

One other thing is that this makes it so that exp() in softmax doesn't get inlined - I talked with @navahgar and I'm going to look into that as a follow up.

ZolotukhinM commented:
> I commented and deleted bc i figured i would just implement it first and then tag, but @ZolotukhinM you saw my comment too quick 😛

I got an e-mail notification and was confused why I couldn't find that comment here lol 😀

Commit:
Update on "[NNC] Dont inlinine intermediates w/multiple uses on cpu"


Differential Revision: [D25688271](https://our.internmc.facebook.com/intern/diff/D25688271)

[ghstack-poisoned]

Commit:
Update on "[NNC] Dont inlinine intermediates w/multiple uses on cpu"


Differential Revision: [D25688271](https://our.internmc.facebook.com/intern/diff/D25688271)

[ghstack-poisoned]
